libraries/chain/block_state.cpp

```diff
@@ -215,6 +223,7 @@ void block_state::verify_qc(const valid_quorum_certificate& qc) const {
    // utility to accumulate voted weights
    auto weights = [&] ( const vote_bitset& votes_bitset ) -> uint64_t {
       uint64_t sum = 0;
+      assert(num_finalizers == votes_bitset.size());
```

Comment:
To demonstrate that the `std::min` on the next line is not needed.

Reply:
> Do we reject the QC pulled from a block if the size is not the same as the number of finalizers for the policy it refers to?
> 
> I want to make sure a bad block wouldn't trip this assertion and would instead trip an earlier check that would throw and exception an reject the block.

There is no earlier check for this that would reject the block. This finction here (`block_state::verify_qc`) is the only check we do on the QC itself.

I added an assert which is enabled only on debug builds, so it doesn't change spring's behavior. Currently because of the `std::min` on line 227 we will accept the QC, but may not consider all votes if there is a mismatch in size. It did seem odd to me that we would quiely accept a size mismatch.

According to what you write, it sounds like  we should change the `assert` into a `EOS_ASSERT` so the block is rejected.

Reply:
Yes, it appears instead of this `assert` we need `EOS_ASSERT`s in `verify_qc` to verify that both the `strong_votes` and `weak_votes` if present are of the right size (they match the number of finalizers) otherwise we reject the QC as invalid. And there should be test cases to demonstrate that check works.

Reply:
Created issue #334 (because of the test cases)